components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71507PhenoTips3423DataMigration.java

```diff
@@ -0,0 +1,558 @@
+        if (pedigreeSettings != null) {
+            newPedigree.put(PEDIGREE_SETTINGS_KEY, pedigreeSettings);
+        }
+
+        // Note1: the assumption is that pedigree is consistent and correct, if a required key
+        //        is missing an exception will be thrown (e.g. by getJSONArray()), which is expected
+        //
+        // Note2: migrated pedigree will preserve all node IDs, so no ID mapping is required
+
+        JSONArray oldData = oldPedigreeData.getJSONArray(OLD_PEDIGREE_GRAPH_KEY);
+
+        // ranks and positions are stored as arrays of integer values indexed by nodeID
+        JSONArray ranks = oldPedigreeData.getJSONArray(OLD_PEDIGREE_RANKS_KEY);
+        JSONArray positions = oldPedigreeData.getJSONArray(OLD_PEDIGREE_POSITIONS_KEY);
+
+        // convert orders as stored to an array of integer order values indexed by nodeID
+        Map<Integer, Integer> order = this.convertOrders(oldPedigreeData.getJSONArray(OLD_PEDIGREE_ORDER_KEY));
+
+        // 1. collect all data in a way addressable by nodeID
+        Map<Integer, JSONObject> data = new HashMap<>();
+        for (Object node : oldData) {
+            JSONObject nodeObj = (JSONObject) node;
+            data.put(nodeObj.getInt("id"), nodeObj);
+        }
+
+        // 2. process all person nodes, collecting data about relationships
+        //    (each node only has outedges specified, so in order to know relationship members need
+        //    to process person nodes first)
+
+        // for each relationship indicated by its nodeID -> a list of paths to the partners involved
+        Map<Integer, List<PartnerPath>> relationshipParents = new HashMap<>();
+
+        for (Map.Entry<Integer, JSONObject> entry : data.entrySet()) {
+            Integer id = entry.getKey();
+            JSONObject nodeObj = entry.getValue();
+
+            if (!nodeObj.has(OLD_PEDIGREE_NODE_TYPE_RELATION)
+                && !nodeObj.has(OLD_PEDIGREE_NODE_TYPE_CHILDHUB)
+                && !nodeObj.has(OLD_PEDIGREE_NODE_TYPE_VIRTUAL)) {
+
+                this.processPerson(id, newPedigree, data, ranks, order, positions);
+
+                // collect data about all relationships this person is part of
+                // (the data is only available via person nodes)
+                this.collectRelationshipData(id, data, relationshipParents);
+            }
+        }
+
+        // 3. process all relationships, now that we know all the partners
+        if (!relationshipParents.isEmpty()) {
+            newPedigree.put(PEDIGREE_RELATIONS_KEY, new JSONArray());
+        }
+        for (Map.Entry<Integer, List<PartnerPath>> entry : relationshipParents.entrySet()) {
+            Integer id = entry.getKey();
+            List<PartnerPath> parents = entry.getValue();
+            if (parents.size() != 2) {
+                throw new IllegalArgumentException("Unsupported pedigree: a relationship with ID="
+                    + id.toString() + "doesn't have exactly two partners");
```

Comment:
need a space between " and d, otherwise exception message will be hard to read, id is merging with text

Reply:
I agree, but this is a one-time message (during data migration) which will likely never happen (for it to happen, need to have an invalid pedigree to start with), so to avoid further delay with merging I will commit as is